packages/@yoda/bluetooth/a2dp.js

```diff
@@ -0,0 +1,584 @@
+          /**
+           * When bluetooth's audio stream state is changed, such as playing/paused.
+           * @event module:@yoda/bluetooth.BluetoothA2dp#audio_state_changed
+           */
+          this.emit('audio_state_changed', protocol.A2DP_MODE.SINK, protocol.AUDIO_STATE.PLAYING)
+        } else if (msg.play_state === 'stopped') {
+          if (this.lastCmd === 'pause') {
+            this.emit('audio_state_changed', protocol.A2DP_MODE.SINK, protocol.AUDIO_STATE.PAUSED)
+          } else {
+            this.emit('audio_state_changed', protocol.A2DP_MODE.SINK, protocol.AUDIO_STATE.STOPPED)
+          }
+          this.lastCmd = null
+        }
+      } else if (msg.a2dpstate === 'opened' && msg.connect_state === 'disconnected') {
+        if (!this.unique) {
+          this.emit('connection_state_changed', protocol.A2DP_MODE.SINK, protocol.CONNECTION_STATE.DISCONNECTED)
+        }
+      }
+
+      if (msg.a2dpstate === 'opened' && msg.connect_state === 'invalid' &&
+        msg.play_state === 'invalid' && msg.broadcast_state === 'opened') {
+        /**
+         * When bluetooth's discovery state is changed, such as undiscoverable or device list changed.
+         * @event module:@yoda/bluetooth.BluetoothA2dp#discovery_state_changed
+         */
+        this.emit('discovery_state_changed', protocol.A2DP_MODE.SINK, protocol.DISCOVERY_STATE.ON)
+      } else if (msg.broadcast_state === 'closed' && this.lastMsg.broadcast_state === 'opened') {
+        this.emit('discovery_state_changed', protocol.A2DP_MODE.SINK, protocol.DISCOVERY_STATE.OFF)
+      }
+      this.lastMsg = Object.assign(this.lastMsg, msg)
+    } else if (msg.action === 'volumechange') {
+      var vol = msg.value
+      if (vol === undefined) {
+        vol = AudioManager.getVolume(AudioManager.STREAM_PLAYBACK)
+      }
+      AudioManager.setVolume(vol)
+      logger.info(`Set volume ${vol} for bluetooth a2dp sink.`)
+      this.emit('audio_state_changed', protocol.A2DP_MODE.SINK, protocol.AUDIO_STATE.VOLUMN_CHANGED, {volumn: vol})
+    }
+  } catch (err) {
+    logger.error(`on a2dpSinkEvent error(${JSON.stringify(err)})`)
+  }
+}
+
+/**
+ * @private
+ */
+BluetoothA2dp.prototype._onSourceEvent = function (data) {
+  if (this._end) {
+    logger.warn('Already destroied!')
+    return
+  }
+  try {
+    var msg = JSON.parse(data[0] + '')
+    logger.debug(`on source event(action:${msg.action})`)
+
+    if (msg.action === 'stateupdate') {
+      logger.debug(
+        `a2dp:${this.lastMsg.a2dpstate}=>${msg.a2dpstate}, conn:${this.lastMsg.connect_state}=>${msg.connect_state}, 
+        bc:${this.lastMsg.broadcast_state}=>${msg.broadcast_state}`)
+      if (msg.a2dpstate === this.lastMsg.a2dpstate && msg.connect_state === this.lastMsg.connect_state &&
+        msg.broadcast_state === this.lastMsg.broadcast_state) {
+        logger.warn('Ignore last source same msg!')
+        return
+      }
+
+      if (msg.a2dpstate === 'opened' && msg.connect_state === 'invalid') {

```

Comment:
Could these if-else be optimized to a state map?

Reply:
Has been changed to the state mapping method as you expected.